File: crabzilla-core/src/main/java/io/github/crabzilla/core/Command.java
package io.github.crabzilla.core;

import com.fasterxml.jackson.annotation.JsonTypeInfo;

import java.io.Serializable;
import java.util.UUID;

@JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="@class")
public interface Command extends Serializable {

  UUID getCommandId();

}

File: crabzilla-core/src/main/java/io/github/crabzilla/core/Command.kt
package io.github.crabzilla.core

import kotlinx.serialization.PolymorphicSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
abstract class Command {
  companion object {
    private val serDer = PolymorphicSerializer(Command::class)
    fun <C : Command> fromJson(json: Json, asJson: String): C {
      return json.decodeFromString(serDer, asJson) as C
    }
  }
  fun toJson(json: Json): String {
    return json.encodeToString(serDer, this)
  }
}
